doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+
+}).listen(port, () => {
+  // Even though all callbacks passed to .listen() are wrapped in a nextTick()
+  // the callback itself exists because the call to the server's .listen()
+  // was made. So the return value would be the id of the server.
+  async_hooks.triggerId();
+});
+```
+
+## Embedder API
+
+Library developers that handle their own I/O will need to hook into the
+AsyncWrap API so that all the appropriate callbacks are called. To accommodate
+this a JavaScript API is provided.
+
+### `class AsyncEvent()`
+
+The class `AsyncEvent` was designed to be extended by the embedder's async
+resources. Using this users can easily trigger the lifetime events of their
+own resources.
+
+The `init()` hook will trigger when an `AsyncEvent` is instantiated.
+ 
+It is important that before/after calls are unwound
+in the same order they are called. Otherwise an unrecoverable exception
+will be made.
+
+```js
+// AsyncEvent() is meant to be extended. Instantiating a new AsyncEvent() also
+// triggers init(). If triggerId is omitted then currentId() is used.
+const asyncEvent = new AsyncEvent(type[, triggerId]);
+
+// Call before() hooks.
+asyncEvent.emitBefore();
+
+// Call after() hooks.
+asyncEvent.emitAfter();
+
+// Call destroy() hooks.
+asyncEvent.emitDestroy();
+
+// Return the unique id assigned to the AsyncEvent instance.
+asyncEvent.asyncId();
+
+// Return the trigger id for the AsyncEvent instance.
+asyncEvent.triggerId();
+```
+
+#### `AsyncEvent(type[, triggerId])`
+
+* arguments
+  * `type` {string} the type of ascycn event
+  * `triggerId` {number} the id of the execution context that created this async
+    event
+* Returns {AsyncEvent} A reference to `asyncHook`.
+
+Example usage:
+
+```js
+class DBQuery extends AsyncEvent {
+  constructor(db) {
+    this.db = db;
+  }
+
+  getInfo(query, callback) {
+    this.db.get(query, (err, data) => {
+      this.emitBefore();
+      callback(err, data)
+      this.emitAfter();
+    });
+  }
+
+  close() {
+    this.db = null;
+    this.emitDestroy();
+  }
+}
+```
+
+#### `asyncEvent.emitBefore()`
+
+* Returns {undefined}
+
+Call all `before()` hooks and let them know a new asynchronous execution
+context is being entered. If nested calls to `emitBefore()` are made the stack
+of `id`s will be tracked and properly unwound.
+
+#### `asyncEvent.emitAfter()`
+
+* Returns {undefined}
+
+Call all `after()` hooks. If nested calls to `emitBefore()` were made then make
+sure the stack is unwound properly. Otherwise an error will be thrown.
+
+If the user's callback throws an exception then `emitAfter()` will
+automatically be called for all `id`'s on the stack if the error is handled by
+a domain or `'uncaughtException'` handler. So there is no need to guard against
+this.
+
+#### `asyncEvent.emitDestroy()`
+
+* Returns {undefined}
+
+Call all `destroy()` hooks. This should only ever be called once. An error will
+be thrown if it is called more than once. This **must** be manually called. If
```

Comment:
Some of our resource objects are destroyed automatically by the gc. Since they are implemented in C++ they actually call the `destroy` callbacks appropriately.